src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/OverloadMethodsDeclarationOrderCheck.java

```diff
@@ -126,27 +210,139 @@ public void visitToken(DetailAST ast) {
      */
     private void checkOverloadMethodsGrouping(DetailAST objectBlock) {
         final int allowedDistance = 1;
+        final Map<String, IndexAndLine> knownMethods = new HashMap<>();
         DetailAST currentToken = objectBlock.getFirstChild();
-        final Map<String, Integer> methodIndexMap = new HashMap<>();
-        final Map<String, Integer> methodLineNumberMap = new HashMap<>();
+
         int currentIndex = 0;
         while (currentToken != null) {
             if (currentToken.getType() == TokenTypes.METHOD_DEF) {
                 currentIndex++;
-                final String methodName =
-                        currentToken.findFirstToken(TokenTypes.IDENT).getText();
-                final Integer previousIndex = methodIndexMap.get(methodName);
-                if (previousIndex != null && currentIndex - previousIndex > allowedDistance) {
-                    final int previousLineWithOverloadMethod =
-                            methodLineNumberMap.get(methodName);
-                    log(currentToken, MSG_KEY,
-                            previousLineWithOverloadMethod);
+                final String methodKey = getMethodDescriptor(currentToken);
+                final IndexAndLine indexAndLine =
+                    new IndexAndLine(currentIndex, currentToken.getLineNo());
+                final IndexAndLine previousMeta = knownMethods.put(methodKey, indexAndLine);
+                if (previousMeta != null
+                    && currentIndex - previousMeta.getIndex() > allowedDistance) {
+                    final int previousLineWithOverloadMethod = previousMeta.getLineNo();
+                    log(currentToken, MSG_KEY, previousLineWithOverloadMethod);
                 }
-                methodIndexMap.put(methodName, currentIndex);
-                methodLineNumberMap.put(methodName, currentToken.getLineNo());
             }
             currentToken = currentToken.getNextSibling();
         }
     }
 
+    /**
+     * Get the first matching {@code modifierGroup} for the modifiers on the method
+     * and combine it with the method's name (separated by {@code SEPARATOR}) to create
+     * a unique identifier for the pair.
+     *
+     * @param currentToken assumed to be a method
+     * @return string that uniquely represents the {@code modifierGroup} and method name
+     */
+    private String getMethodDescriptor(DetailAST currentToken) {
```

Comment:
Split this into 2 methods:
`findModifierGroupPattern` - it just returns the matched pattern or the default.
I am still looking over the main method for naming and coding.

Reply:
I think if we only return index splitting into two methods is no longer required? If you still want two methods, what would the second method do?